### PR TITLE
Feature/farm 419 profile repository

### PR DESCRIPTION
**Pull request by LeeMatthewHiggins**
[FARM-419 - profile repository]

#### 📲 What

Added profile switching.
Added profile stats for active crops
Ensured Mock repos have a dataset per profile
Reload app on profile switch
Split out Recommendations engine weights to own repo

#### 🤔 Why
		
Needed to build business logic layer for profile switching
		
#### 🛠 How
		
Currently uses only Mock repos
Firebase TODO

#### 🕵️‍♂️ How to test

Navigate to profile screen, switch profile, note name change.
Add crops to profiles and ensure data consistent between profiles.

#### Reviewers

@farmsmart/amido @farmsmart/wamf

### CODE REVIEW

File: lib/model/bloc/plot/PlotDetailProvider.dart [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import 'dart:async';
 
-import 'package:farmsmart_flutter/model/model/NewStageEntity.dart';
+import 'package:farmsmart_flutter/model/model/StageEntity.dart';
```
Comment (davidfranquet) [inline lib/model/bloc/plot/PlotDetailProvider.dart:3]: 👍

File: lib/model/bloc/plot/PlotStatistics.dart [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import 'StageBusinessLogic.dart';
+
+class PlotStatistics {
+
+  StageBusinessLogic _stageLogic = StageBusinessLogic();
+
+  int compeletedCount(List<PlotEntity> plots) {
+    int completed = 0;
+    plots.forEach((plot){
+      completed += _stageLogic.isComplete(plot.stages.last)? 1:0;
```
Comment (davidfranquet) [inline lib/model/bloc/plot/PlotStatistics.dart:12]: Magic number
```suggestion
      completed += _stageLogic.isComplete(plot.stages.last)? 1 : 0;
```

File: lib/model/bloc/plot/PlotStatistics.dart [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    });
+    return completed;
+  }
+
+  int activeCount(List<PlotEntity> plots) {
+     int completed = 0;
+    plots.forEach((plot){
+      bool active = _stageLogic.isStarted(plot.stages.first) && !_stageLogic.isComplete(plot.stages.last);
+      completed += active ? 1:0;
```
Comment (davidfranquet) [inline lib/model/bloc/plot/PlotStatistics.dart:21]: 🔮 number
```suggestion
      completed += active ? 1 : 0;
```

File: lib/ui/profile/Profile.dart [hunk 4]
```diff
@@ -146,35 +156,44 @@ class _DefaultStyle extends UserProfileStyle {
...
 
-class UserProfile extends StatelessWidget {
-  final UserProfileViewModel _viewModel;
-  final UserProfileStyle _style;
+class Profile extends StatelessWidget {
+  final ViewModelProvider<ProfileViewModel> _viewModelProvider;
+  final ProfileStyle _style;
 
-  const UserProfile({
+  const Profile({
```
Comment (davidfranquet) [inline lib/ui/profile/Profile.dart:165]: 👍

File: lib/ui/profile/Profile.dart [hunk 5]
```diff
@@ -250,16 +269,16 @@ class UserProfile extends StatelessWidget {
...
       crossAxisAlignment: CrossAxisAlignment.start,
       children: <Widget>[
         Text(
-          _viewModel.activeCrops.toString(),
+          viewModel.activeCrops.toString(),
           style: _style.subtitleTextStyle,
         ),
         Text(
-          _LocalisedStrings.activeCrops(),
+          Intl.message(_LocalisedStrings.activeCrops()),
```
Comment (davidfranquet) [inline lib/ui/profile/Profile.dart]: _LocalisedStrings.activeCrops() is already an Intl.message :)

```suggestion
          _LocalisedStrings.activeCrops(),
```

File: lib/ui/profile/Profile.dart [hunk 6]
```diff
@@ -268,16 +287,16 @@ class UserProfile extends StatelessWidget {
...
       crossAxisAlignment: CrossAxisAlignment.start,
       children: <Widget>[
         Text(
-          _viewModel.completedCrops.toString(),
+          viewModel.completedCrops.toString(),
           style: _style.subtitleTextStyle,
         ),
         Text(
-          _LocalisedStrings.completedCrops(),
+          Intl.message(_LocalisedStrings.completedCrops()),
```
Comment (davidfranquet) [inline lib/ui/profile/Profile.dart]: Same here
```suggestion
          _LocalisedStrings.completedCrops(),
```

File: lib/ui/profile/Profile.dart [hunk 7]
```diff
@@ -286,15 +305,16 @@ class UserProfile extends StatelessWidget {
...
+  Widget _buildButton(BuildContext context, ProfileViewModel viewModel) {
     return Padding(
       padding: _Constants.edgePadding,
       child: Column(
         children: <Widget>[
           RoundedButton(
             viewModel: RoundedButtonViewModel(
-              title: _LocalisedStrings.buttonTitle(),
-              onTap: () => _viewModel.switchProfileAction(),
+              title: Intl.message(_LocalisedStrings.buttonTitle()),
```
Comment (davidfranquet) [inline lib/ui/profile/Profile.dart]: Same
```suggestion
              title: _LocalisedStrings.buttonTitle(),
```

File: lib/ui/profitloss/RecordTransactionListItem.dart [hunk 8]
```diff
@@ -27,7 +27,7 @@ class _LocalisedStrings {
 }
 
 class _Strings {
-  static final emptyString = "";
+  static final emptyString = Intl.message("");
```
Comment (davidfranquet) [inline lib/ui/profitloss/RecordTransactionListItem.dart]: I think that we don't need to localise an empty string :)
```suggestion
  static final emptyString = '';
```

**Review by davidfranquet — CHANGES_REQUESTED** [review]
Great work! Just minor comments and some wrong Intl.message()

File: lib/model/bloc/plot/PlotStatistics.dart (hunk 2 above)
Comment (LeeMatthewHiggins) [reply lib/model/bloc/plot/PlotStatistics.dart:12]: I think this is ok here and its a literal constant of 1. we're counting..

File: lib/ui/profile/Profile.dart (hunk 5 above)
Comment (LeeMatthewHiggins) [reply lib/ui/profile/Profile.dart]: ah missed this in merge

File: lib/model/bloc/plot/PlotStatistics.dart (hunk 2 above)
Comment (davidfranquet) [reply lib/model/bloc/plot/PlotStatistics.dart:12]: 👌

File: lib/ui/profile/Profile.dart (hunk 5 above)
Comment (davidfranquet) [reply lib/ui/profile/Profile.dart]: np!

Comment (LeeMatthewHiggins) [reply lib/ui/profile/Profile.dart]: 👍

File: lib/ui/profile/Profile.dart (hunk 6 above)
Comment (LeeMatthewHiggins) [reply lib/ui/profile/Profile.dart]: 👍

File: lib/ui/profile/Profile.dart (hunk 7 above)
Comment (LeeMatthewHiggins) [reply lib/ui/profile/Profile.dart]: 👍

File: lib/ui/profitloss/RecordTransactionListItem.dart (hunk 8 above)
Comment (LeeMatthewHiggins) [reply lib/ui/profitloss/RecordTransactionListItem.dart]: 👍

**Review by davidfranquet — APPROVED** [review]
LGTM! 🚀

File: lib/model/bloc/profile/ProfileDetailProvider.dart [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+      _snapshot = transform(from: null);
+      _snapshot.refresh();
+    }
+    return _snapshot;
+  }
+
+  @override
+  ProfileViewModel transform({ProfileEntity from}) {
+    List<ProfileListItemViewModel> list = []; //TODO; replace with real items
+    for (var i = 0; i < 8; i++) {
```
Comment (albertmiro) [inline lib/model/bloc/profile/ProfileDetailProvider.dart:72]: Magic number? 8

File: lib/model/bloc/profile/ProfileDetailProvider.dart [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+      _snapshot = transform(from: null);
+      _snapshot.refresh();
+    }
+    return _snapshot;
+  }
+
+  @override
+  ProfileViewModel transform({ProfileEntity from}) {
+    List<ProfileListItemViewModel> list = []; //TODO; replace with real items
```
Comment (albertmiro) [inline lib/model/bloc/profile/ProfileDetailProvider.dart:71]: That would be done in another PR?

File: lib/model/bloc/profile/SwitchProfileListProvider.dart [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+      _currentProfile,
+    );
+    final listItems = profiles.map((profile) {
+      return transformer.transform(from: profile);
+    }).toList();
+    return SwitchProfileListViewModel(
+      title: _LocalisedStrings.switchProfile(),
+      actionTitle: _LocalisedStrings.switchProfile(),
+      items: listItems,
+      addProfileAction: null,
```
Comment (albertmiro) [inline lib/model/bloc/profile/SwitchProfileListProvider.dart:86]: That's ok?

File: lib/ui/mockData/MockUserProfileViewModel.dart [hunk 12]
```diff
@@ -45,18 +48,18 @@ class MockUserProfileViewModel {
 }
 
 class MockUserProfileListItemViewModel {
-  static UserProfileListItemViewModel build(index) {
-    return UserProfileListItemViewModel(
-      title: _mockActionTitle[index],
+  static ProfileListItemViewModel build(index) {
+    return ProfileListItemViewModel(
+      title: Intl.message(_mockActionTitle[index]),
```
Comment (albertmiro) [inline lib/ui/mockData/MockUserProfileViewModel.dart:53]: `Intl.message` should be removed, it's in mock area, probably you got it from rebase.

File: lib/ui/mockData/MockUserProfileViewModel.dart [hunk 13]
```diff
@@ -45,18 +48,18 @@ class MockUserProfileViewModel {
...
       isDestructive: index != 7 ? false : true,
     );
   }
 
-  static UserProfileListItemViewModel buildLarger(index) {
-    return UserProfileListItemViewModel(
-      title: _mockActionTitleLarger[index],
+  static ProfileListItemViewModel buildLarger(index) {
+    return ProfileListItemViewModel(
+      title: Intl.message(_mockActionTitleLarger[index]),
```
Comment (albertmiro) [inline lib/ui/mockData/MockUserProfileViewModel.dart:62]: `Intl.message` should be removed, it's in mock area, probably you got it from rebase.

File: lib/model/bloc/profile/ProfileDetailProvider.dart (hunk 9 above)
Comment (LeeMatthewHiggins) [reply lib/model/bloc/profile/ProfileDetailProvider.dart:72]: this will be replaced with the real implementation, just using the MOCK for now

File: lib/model/bloc/profile/ProfileDetailProvider.dart (hunk 10 above)
Comment (LeeMatthewHiggins) [reply lib/model/bloc/profile/ProfileDetailProvider.dart:71]: yep

File: lib/model/bloc/profile/SwitchProfileListProvider.dart (hunk 11 above)
Comment (LeeMatthewHiggins) [reply lib/model/bloc/profile/SwitchProfileListProvider.dart:86]: for now, I need to chat integration to be able to implement the add profile feature.

File: lib/ui/mockData/MockUserProfileViewModel.dart (hunk 13 above)
Comment (LeeMatthewHiggins) [reply lib/ui/mockData/MockUserProfileViewModel.dart:62]: ah nuts
will remove

Comment (LeeMatthewHiggins) [reply lib/ui/mockData/MockUserProfileViewModel.dart:62]: done

File: lib/ui/mockData/MockUserProfileViewModel.dart (hunk 12 above)
Comment (LeeMatthewHiggins) [reply lib/ui/mockData/MockUserProfileViewModel.dart:53]: done

**Review by albertmiro — APPROVED** [review]
